packages/types/src/pnum.ts

```diff
@@ -0,0 +1,115 @@
+import { BigNumber } from 'bignumber.js';
+import { round } from '@penumbra-zone/types/round';
+import { LoHi, joinLoHi, splitLoHi } from '@penumbra-zone/types/lo-hi';
+import { Amount } from '@penumbra-zone/protobuf/penumbra/core/num/v1/num_pb';
+import { ValueView } from '@penumbra-zone/protobuf/penumbra/core/asset/v1/asset_pb';
+import { getAmount, getDisplayDenomExponentFromValueView } from '@penumbra-zone/getters/value-view';
+import { removeTrailingZeros } from '@penumbra-zone/types/shortify';
+
+/**
+ * pnum (penumbra number)
+ *
+ * In Penumbra a number can be in the form of a base unit (bigint, LoHi, Amount, ValueView)
+ * or a number with decimals for display purposes (string, number)
+ *
+ * This function handles all these cases automatically internally
+ * - when input is a bigint, LoHi, Amount, or ValueView, it is assumed to be in base units
+ * - when input is a string or number, it is multiplied by 10^exponent and converted to base units
+ *
+ * Likewise for all methods, the outputs are
+ * - in base units for bigint, LoHi, Amount and ValueView
+ * - in display form with decimals for string and number
+ *
+ * @param input
+ * @param exponent
+ */
+function pnum(
+  input: string | number | LoHi | bigint | Amount | ValueView | undefined,
+  exponentInput = 0,
+) {
+  let value: BigNumber;
+  let exponent = exponentInput;
+
+  if (typeof input === 'string' || typeof input === 'number') {
+    value = new BigNumber(input).shiftedBy(exponent);
+  } else if (typeof input === 'bigint') {
+    value = new BigNumber(input.toString());
+  } else if (input instanceof ValueView) {
+    const amount = getAmount(input);
+    value = new BigNumber(joinLoHi(amount.lo, amount.hi).toString());
+    exponent =
+      input.valueView.case === 'knownAssetId' ? getDisplayDenomExponentFromValueView(input) : 0;
+  } else if (
+    input instanceof Amount ||
+    (typeof input === 'object' && 'lo' in input && 'hi' in input && input.lo && input.hi)
```

Comment:
Suggestion: Wouldn't this work the same?

```ts
else if (input instanceof Amount) {
    value = new BigNumber(joinLoHi(input.lo, input.hi).toString());
  }
```

Reply:
This case catches an instance of Amount **and LoHi**, in your check a LoHi will not be true.

Reply:
No test appears to fail when removing that line. Can you add a test for coverage for that?

Reply:
Sure